File: learning-content-service/src/main/java/com/rin/learningcontentservice/utils/TextUtils.java
package com.rin.learningcontentservice.utils;

import java.text.Normalizer;

public class TextUtils {
    private TextUtils() {
        /* This utility class should not be instantiated */
    }


    // Kiểm tra token có chứa dấu câu không
    public static boolean hasPunctuation(String token) {
        if (token == null || token.isEmpty()) return false;

        String punctuationMarks = ".,!?;:\"()[]{}…—–-";
        for (char c : token.toCharArray()) {
            if (punctuationMarks.indexOf(c) >= 0) {
                return true;
            }
        }
        return false;
    }
    public static String normalizeWordLower(String word) {
        if (word == null) return null;

        return word.toLowerCase().trim();
    }

    public static String normalizeWordSoft(String word) {
        if (word == null) return null;

        word = word.trim();

        // remove ký tự không hợp lệ ở đầu
        word = word.replaceAll("^[^a-zA-Z'-]+", "");

        // remove ký tự không hợp lệ ở cuối
        word = word.replaceAll("[^a-zA-Z'-]+$", "");

        return word;
    }

    // Tạo slug từ chuỗi (bỏ dấu, thay space bằng -)
    public static String createSlug(String input) {
        if (input == null) return null;

        // Bỏ dấu tiếng Việt
        String normalized = Normalizer.normalize(input, Normalizer.Form.NFD);
        String withoutAccents = normalized.replaceAll("\\p{InCombiningDiacriticalMarks}+", "");

        // Xóa ký tự đặc biệt, thay space bằng -
        String cleaned = withoutAccents.toLowerCase().replaceAll("[^a-z0-9\\s-]", "");
        String dashed = cleaned.trim().replaceAll("\\s+", "-");

        return dashed.replaceAll("-{2,}", "-");
    }
}

File: language-processing-service/src/utils/text_utils.py
# src/utils/text_utils.py
import re
def clean_for_spacy(word: str) -> str:
    """
    Làm sạch dấu câu để phân tích spaCy, nhưng KHÔNG thay đổi database
    Ví dụ: "he." -> "he", "why?" -> "why", "Hello!" -> "Hello"
    """
    if not word:
        return word
    
    # Bỏ dấu câu ở đầu và cuối
    word = re.sub(r'^[^\w\s\'-]+', '', word)
    word = re.sub(r'[^\w\s\'-]+$', '', word)
    return word

def normalize_word_lower(word: str) -> str:
    if word is None:
        return None
    return word.lower().strip()

def normalize_word_soft(word: str) -> str:
    if word is None:
        return None
    word = word.strip()
    word = re.sub(r"^[^a-zA-Z'-]+", "", word)
    word = re.sub(r"[^a-zA-Z'-]+$", "", word)
    return word

def create_slug(word: str) -> str:
    if word is None:
        return None
    slug = word.lower()
    slug = re.sub(r"[^a-z0-9]+", "-", slug)
    slug = slug.strip("-")
    return slug if slug else None